librosa/core/pitch.py

```diff
@@ -335,3 +338,281 @@ def piptrack(y=None, sr=22050, S=None, n_fft=2048, hop_length=None,
                                   + dskew[idx[:, 0], idx[:, 1]])
 
     return pitches, mags
+
+
+def yin(y, sr=22050, win_length=1024, hop_length=None, fmin=55, fmax=900, peak_threshold=0.1):
+    '''Fundamental frequency (F0) estimation. [1]_
+
+    .. [1] De Cheveigné, A., & Kawahara, H. (2002).
+        "YIN, a fundamental frequency estimator for speech and music."
+        The Journal of the Acoustical Society of America, 111(4), 1917-1930.
+
+     Parameters
+     ----------
+     y : np.ndarray [shape=(n,)]
+         audio time series.
+
+     sr : number > 0 [scalar]
+         sampling rate of `y` in Hertz.
+
+     win_length : int [scalar]
+         length of the window in samples.
+         By default, win_length=1024 corresponds to a time scale of 46 ms at
+         a sampling rate of 22050 Hz.
+
+     hop_length : None or int > 0 [scalar]
+         number of audio samples between adjacent pYIN predictions.
+         If `None`, defaults to `frame_length // 4`.
+
+     fmin: number > 0 [scalar]
+         minimum frequency in Hertz.
+
+     fmax: number > 0 [scalar]
+         maximum frequency in Hertz.
+
+     peak_threshold: number > 0 [scalar]
+         absolute threshold for peak estimation.
+
+     Returns
+     -------
+     f0: np.ndarray [shape=(n_frames,)]
+         time series of fundamental frequencies in Hertz.
+
+     Examples
+     --------
+     Computing a fundamental frequency curve from an audio input:
+     >>> y = librosa.chirp(440, 880, duration=7.0)
+     >>> yin(y)
+     array([441.15076669, 441.6640896 , 442.16219657, ...,
+        871.59470743, 872.63699131, 873.66696896])
+     '''
+    # Set the default hop if it is not already specified.
+    if hop_length is None:
+        hop_length = int(win_length // 4)
+
+    # Check that audio is valid.
+    util.valid_audio(y, mono=True)
+
+    # Autocorrelation.
+    y_frames = util.frame(y, frame_length=2*win_length, hop_length=hop_length)
+    n_frames = y_frames.shape[-1]
+    a = np.fft.fft(y_frames, 2*win_length, axis=0)
+    b = np.fft.fft(y_frames[win_length::-1, :], 2*win_length, axis=0)
+    acf_frames = np.fft.ifft(a*b, 2*win_length, axis=0).real[win_length:]
+
+    # Energy terms.
+    window = scipy.signal.windows.boxcar(win_length + 1)
+    energy_frames = scipy.signal.fftconvolve(
+        y_frames ** 2, window[:, None], mode="valid", axes=0)
```

Comment:
I'm a little confused by this line, can you explain what's going on here?  Why are we convolving a box window along the lag axis?

Reply:
The idea here is to calculate the energy of the signal within a window. This is for the first two terms of equation 7 in the YIN paper. Another way to implement it would be to use `np.cumsum` and take differences.

Reply:
Ah, I see.  cumsum would be faster, though I guess there are wrapping effects to contend with.  Feel free to ignore the previous comment about the window being user-settable for now.

Reply:
I replaced the convolution with the `cumsum` method.